### ISSUE TITLE
_z_update()  needs a small adjustment

### ISSUE THREAD

**Issue by Filgs** [catwell/fakeredis#8]
@catwell ,
Your fakeredis library is exactly what I was looking for.
A nice local Redis 100% in Lua able to run without remote Server.

I just would like to inform you that I am doing persistence on the self.data database table.
Therefore I noticed a small issue in zadd command, when the key was already stored at the database previously.
The solution is very simple, somehow you forgot to add the .s at _z_update() function.
Without the .s you are comparing 2 tables and this generates a Lua error.
With just this simple change it works perfectly well !!!

`local _z_update = function(x, p)`
`  local l = x.list`
`  local found = _z_remove(x, p.v)`
`  local ix = nil`
`  for i=1,#l do`
`..if l[i].s > p.s then   -- @@ Filgs: the .s was missing in the original library`
`....ix = i; break`
`..end`
`end`
`  if not ix then ix = #l+1 end`
`  _z_insert(x, ix, p)`
`  return found`
`end`

Very nice job !
@Filgs

**Comment by catwell** [comment]
This fix is wrong (comparison is a bit more complicated, see how it breaks tests for equal scores) and the issue should not happen because `l[i]` and `p` should have a metatable that defines comparison.

Can you share code that causes the issue please?

**Comment by Filgs** [comment]
@catwell ,
Thanks for your warning. Indeed the comparison is much more complicated and my first approach to just eliminate the Lua interpreter error by adding the ".s" in the comparison was naive.
I made some extra tests and realized that the sequence of the results are not OK as you expected.

The root cause of the failure is exactly the missing metatable (_z_p_mt) which is not assigned to the sorted pairs as you mentioned.

The code I am using to persist the database is quite simple. I am just converting the redis.data table to JSON format and saving to a file. 
`file.saveJSON (redis.data, "redis.json") -- persist database`

Later on when the program restarts I load the JSON file into redis.table.
`redis.data = file.loadJSON ("redis.json") -- load database`

As expected the assigned metatable  (_z_p_mt) is lost during this procedure and caused the failure I got previously.

The good news is that I added a small function to re-assign the metatable  (_z_p_mt) to the sorted sets pairs after loading the file. See below the code:
```
local setZPMT = function (self)  
  for _, v in pairs (self.data) do
    if v.ktype == "zset" then  -- check if key is sorted set
      local n = #v.value.list
      for i = 1, n do setmetatable (v.value.list[i], _z_p_mt) end --reassign metatable to the sorted pairs
    end
  end
end
```

Now I can safely restore the database and the metatables accordingly.
I hope you may add this quite simple function but powerful resource into your library.

**Comment by catwell** [comment]
Thank you for the test case. Yes, such a function is exactly what should be added to the library. I will do it later this week.

**Comment by catwell** [comment]
@Filgs Sorry to ask again, but can you post actual code that breaks? I tried to reproduce the issue but I couldn't and when I read the code it looks to me like *even* if you remove the metatables by serializing and deserializing this way it should still work, because the metatable is always added on the right member by all operations before calling `_z_update`.

So I need an actual code sample that fails. This, for instance, does not:

    R:zadd("foo", 10, "A")
    R.data = json.decode(json.encode(R.data))
    R:zadd("foo", 20, "B")

**Comment by catwell** [comment]
Oooh, I understand the issue now. It actually only occurs with Lua 5.1 or LuaJIT, not 5.2 and 5.3!

This is a change of behavior in Lua. I had to commit bdbda1d6533816a9e97d5563f76f48d4a20e8280 to see it. In 5.1, the comparison metatables use `getcomphandler` so both sides need the metatable. In 5.2 and 5.3, it was changed to `getbinhandler` so only one side needs it. Tricky!

**Comment by catwell** [comment]
I fixed the issue by stopping to use metatables at all. I released version 0.5 which has the fix. There is no need for a specific function anymore.

**Comment by Filgs** [comment]
@catwell ,
Your assumptions are correct. 
Indeed I am using Lua 5.1 and the idea to avoid using metatable for Zset pairs is much better than restoring the metatables after database reload by a function like the suggested setZPMT().
The function _z_pair_lt() does the job effectively.

I have tested the new version 0.5 and it works 100% fine.
Thanks a lot for your support!
